### ISSUE TITLE
[Feature Request] Provide syntactic sugar to make programming in a functional style easier

### ISSUE THREAD

**Issue by astares** [guillep/pharo-core#11587]
At ESUG 2022 there was discussed about syntactic sugar to make programming in a functional style easier. 
Eliminating the ().

See https://twitter.com/StOnSoftware/status/1562411468901031937

![image](https://user-images.githubusercontent.com/5980033/186508723-1a73b9cb-3cd5-4fe2-9d0f-1b919cfd933b.png)

![image](https://user-images.githubusercontent.com/5980033/186508797-7191fa12-5dec-4367-bd49-d5ce66ebc5d0.png)

Attila Magyar suggested on Twitter that this can be implemented as a proxy with DNU handler, without adding new syntax.
See  https://twitter.com/zeroflag/status/1562414140639137792

![image](https://user-images.githubusercontent.com/5980033/186508520-7a208df3-a751-436c-ba57-68b73a2cca97.png)

We could provide such a feature in the standard image if people need and upvote this.

**Comment by zeroflag** [comment]
@astares thanks for sharing. Here is a POC implementation of this: https://gist.github.com/zeroflag/1634116227c3b811f0692db4f2889634

<blockquote><img src="https://github.githubassets.com/images/modules/gists/gist-og-image.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> Gist</div><div><strong><a href="https://gist.github.com/zeroflag/1634116227c3b811f0692db4f2889634">chain.st</a></strong></div><div>GitHub Gist: instantly share code, notes, and snippets.</div></blockquote>

**Comment by LucFabresse** [comment]
using DNU for this is not that good because it requires 2 lookups for each chained message.

**Comment by guillep** [comment]
I think from all the extensions shown, destructuring assignments are the easiest to integrate:
 - less parser friction, it should be implemented without breaking the parser
 - backwards compatible, it relies on syntax that is currently invalid so it will not break existing programs
 - no extra runtime cost (just syntax sugar for collection accesses)

It would be nice that somebody writes a Phep on this (or in the more complex one).
The slides can be a starting point!

**Comment by dvmason** [comment]
> @astares thanks for sharing. Here is a POC implementation of this: https://gist.github.com/zeroflag/1634116227c3b811f0692db4f2889634
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/e249f005c9359542c16b3dc40dbe3e950c23c49a86601d5d2461a59834675446/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f696d616765732f6d6f64756c65732f67697374732f676973742d6f672d696d6167652e706e67"><img alt="" height="14" src="https://camo.githubusercontent.com/b6a12909f1e31185a69a73d59208c507a992236d3230f9fc18e85058ae3d19e7/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f66617669636f6e732f66617669636f6e2e737667"> Gist**[chain.st](https://gist.github.com/zeroflag/1634116227c3b811f0692db4f2889634)**GitHub Gist: instantly share code, notes, and snippets.

Yes, if you look at the https://github.com/dvmason/Pharo-Functional repo, you will see that we already have that as well, but as @LucFabresse mentions, this requires a DNU and `perform:` for each message, neither of which are super-fast operations!

<blockquote><img src="https://github.githubassets.com/images/modules/gists/gist-og-image.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> Gist</div><div><strong><a href="https://gist.github.com/zeroflag/1634116227c3b811f0692db4f2889634">chain.st</a></strong></div><div>GitHub Gist: instantly share code, notes, and snippets.</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/33ca10ea5711160ed86e8039d9fa77f51e21f304b4fa4cae8c91f85a465ff0ae/dvmason/Pharo-Functional" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/dvmason/Pharo-Functional">GitHub - dvmason/Pharo-Functional: Functional support for Pharo</a></strong></div><div>Functional support for Pharo. Contribute to dvmason/Pharo-Functional development by creating an account on GitHub.</div></blockquote>

**Comment by dvmason** [comment]
> I think from all the extensions shown, destructuring assignments are the easiest to integrate:
> 
> * less parser friction, it should be implemented without breaking the parser
> * backwards compatible, it relies on syntax that is currently invalid so it will not break existing programs
> * no extra runtime cost (just syntax sugar for collection accesses)
> 
> It would be nice that somebody writes a Phep on this (or in the more complex one). The slides can be a starting point!

All of the syntax proposals are backwards compatible... they all are invalid syntax in a standard parser. And all the code and implementation is in the repo: https://github.com/dvmason/Pharo-Functional

I would argue that of the proposals, the parrot operator and collection literals are the easiest, zero-cost options. Destructuring assignments also have the same cost as writing them out.

Initializing locals at point of declaration is also zero-cost and I'd love to see them.

Expressions as messages are more of a mental shift. They can make for some very elegant code, but they aren't quite zero-cost, and the more elegant code may be a matter of taste.
